runtime/memprof.c

```diff
@@ -652,39 +712,37 @@ void caml_memprof_track_young(tag_t tag, uintnat wosize, int from_caml)
     caml_gc_dispatch();
   }
 
-  /* This condition happens either in the case of an exception or if
-     the callback returned [None]. If these cases happen frequently,
-     then we need to call [flush_deleted] somewhere to prevent a
-     leak. */
-  if (t_idx == Invalid_index)
-    flush_deleted();
-
-  caml_raise_if_exception(res);
-
-  /* Re-allocate the block in the minor heap. We should not call the
+  /* Re-allocate the blocks in the minor heap. We should not call the
      GC after this. */
   Caml_state->young_ptr -= whsize;
 
   /* Make sure this block is not going to be sampled again. */
   shift_sample(whsize);
 
-  if (t_idx != Invalid_index) {
-    /* If the execution of the callback has succeeded, then we start the
-       tracking of this block..
-
-       Subtlety: we are actually writing [t->block] with an invalid
-       (uninitialized) block. This is correct because the allocation
-       and initialization happens right after returning from
-       [caml_memprof_track_young]. */
-    trackst.entries[t_idx].block = Val_hp(Caml_state->young_ptr);
-    CAMLassert(trackst.entries[t_idx].cb_alloc_called);
-    if (t_idx < trackst.young) trackst.young = t_idx;
+  for (i = 0; i < allocs_sampled; i++) {
+    struct sampled_young_alloc* s = &sampled_allocs[i];
+    res = allocs_sampled == 1 ? cb_res : Field(cb_res, i);
+    if (res != Val_unit) {
+      /* If the execution of the callback has succeeded, then we start the
+         tracking of this block..
+
+         Subtlety: we are actually writing [t->block] with an invalid
+         (uninitialized) block. This is correct because the allocation
+         and initialization happens right after returning from
+         [caml_memprof_track_young]. */
+      new_tracked(s->n_samples, Make_header(s->wosize, tag, Caml_white),
```

Comment:
There is a slight issue here: it is possible that the `new_tracked` call fails (in case of OOM). Then, this means we called the allocation callback, but will never call the promotion or deallocation callback, which can lead to a memory leak in the client code.

This is perhaps not too important, because this happens only in case of anOOM condition, which will likely crash the program soon anyway, but if we keep this implementation, I would prefer crashing in `new_tracked` directly instead of leaving the program in a somewhat inconsistent state.